### DISCUSSION TITLE
Preprocess punctuation

### DISCUSSION THREAD

**Discussion by mortii** [mortii/anki-morphs#273] [Ideas]
> It introduces another problem, though, namely that with `Ankimorphs: Simple Space Splitter` punctuation becomes an integral part of the morph. For example, "است.",‎ "است،"‎, "است؟"‎, "است!" and "است" in Persian are all considered distinct morphs, while it would be best if they were unified as the same morph without the punctuation ("است"). Can’t we just remove a given set of characters when processing the field, using a regex or some other method? [This page](https://www.fileformat.info/info/unicode/category/index.htm) has lists of punctuation marks, but they have way too many symbols for practical use with Ankimorphs; for example, the ‘[Punctuation, Other](https://www.fileformat.info/info/unicode/category/Po/index.htm)’ category has no less than 628 characters, the vast majority of which aren’t likely to ever be used with Ankimorphs. I suggest adding a configurable list of punctuation characters (as a textbox/string), with reasonable defaults (including Latin, Greek, Armenian, Hebrew, Arabic and Devanagari for example, but excluding Mandaic). This way the list of punctuation characters won’t be too complex and will accommodate most users (and as the list is configurable, if someone *does* learn Mandaic with Ankimorphs, they can add the relevant characters easily…).

_Originally posted by @rwmpelstilzchen in https://github.com/mortii/anki-morphs/issues/250#issuecomment-2266154888_

**Comment by mortii** [comment]
Adding a new preprocess option to remove specified characters could definitely help with this. It would be a global option though, so you couldn't attach it to any specific note filter, but if that becomes a problem then using different anki profiles might be a sufficient work around.

**Reply by rwmpelstilzchen** [reply]
Moreover, if I understand correctly this solution will eliminate the need to have two basic morphemizers (`AnkiMorphs: Simple Space Splitter` and `AnkiMorphs: SSS + Punctuations`), which will make things clearer UI/UX-wise.

**Reply by mortii** [reply]
> this solution will eliminate the need to have two basic morphemizers

The the `SSS + Punctuation` preserves words like `mother-in-law` and `quelqu'un`. You could remove the hyphens and apostrophes I guess, but that wouldn't be ideal, especially if you are just starting to learn the language.

**Reply by rwmpelstilzchen** [reply]
Oh, I see.

**Reply by mortii** [reply]
@rwmpelstilzchen actually, you are completely right--those words are preserved with the SSS morphemizer as well. If we can make this option work, then we should definitely just merge them/remove one.

**Reply by buqamura** [reply]
Hi, I'd just like to say thanks for working on this ... and to confirm that the new morphemizer is definitely an issue for the Levantine Arabic deck I'm working with right now: the final word of any question is categorized differently, leading to the need to click "K" when those "doubles" come up, which obviously isn't ideal for the deck stats -- especially with the relatively small deck that I'm working with.

I'll be looking forward to the solution; let me know if I can help with testing etc.

**Reply by rwmpelstilzchen** [reply]
\<OT\>
Hi @buqamura, I see you don’t have any contact information on GitHub, so I venture writing here an off-topic message…
If it’s *South* Levantine Arabic you are learning, [here](https://me.digitalwords.net/tmp/adi/ajp.apkg) is a deck you might find useful. It consists of cards from [this deck](https://ankiweb.net/shared/info/1000027551) and [this deck](https://ankiweb.net/shared/info/515014304) from AnkiWeb, as well as [this one](https://lingualism.com/product/palestinian-colloquial-arabic-vocabulary-anki-flashcards/) and [this one](https://lingualism.com/product/palestinian-arabic-verbs-anki-flashcards/) from [Lingualism](https://lingualism.com/), all combined into one deck with one note type. The last two sources are not free; if you use the combined deck please purchase them: a great deal of work has been done creating them, and the creators deserve being paid. I made the combined deck before switching to AnkiMorphs from MorphMan, so you might want to remove the old tags and fields.
Apologies for the interruption… \^\_\^;
\</OT\>

**Reply by mortii** [reply]
@buqamura sorry to hear that. 

Another potential work around for this is to use a similar approach detailed in the ["morphs don't split correctly" section](https://mortii.github.io/anki-morphs/user_guide/known-problems.html): basically substitute the `<br>` part with whatever characters is causing the "duplicate" morphs, e.g., question marks. It's going to make the text look weird, but that might be preferable. 

Unfortunately, fixing this problem might break backwards compatibility, so it won't be done immediately. Realistically, it will take around 2-3 months.

**Comment by mortii** [comment]
This is what I have in mind:

![Screenshot from 2024-09-12 12-04-03](https://github.com/user-attachments/assets/8288140b-1bf9-4a73-976d-d63377c851c0)

where any character entered into the field will either be stripped from the text or ignored, depending which ever approach turns out to me more efficient. Does that seem reasonable?

**Comment by mortii** [comment]
Here is the testing version, which should be fully working: https://drive.google.com/file/d/1fEpsXXq4l3QsGoI3C8mbWq-nVAKur5a5/view?usp=sharing

It shows up as "ankimorphs" in the add-ons window. Disable the "AnkiMorphs" add-on while using it.

We can remove the morphemizers later, which means this doesn't break backwards compatibility and can therefore be released at any time.

**Comment by mortii** [comment]
Released in [v3.2.0](https://github.com/mortii/anki-morphs/releases/tag/v3.2.0)
